hoot-core/src/main/cpp/hoot/core/io/OsmApiChangesetElement.cpp

```diff
@@ -0,0 +1,195 @@
+/*
+ * This file is part of Hootenanny.
+ *
+ * Hootenanny is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * --------------------------------------------------------------------
+ *
+ * The following copyright notices are generated automatically. If you
+ * have a new notice to add, please use the format:
+ * " * @copyright Copyright ..."
+ * This will properly maintain the copyright information. DigitalGlobe
+ * copyrights will be updated automatically.
+ *
+ * @copyright Copyright (C) 2018 DigitalGlobe (http://www.digitalglobe.com/)
+ */
+
+#include "OsmApiChangesetElement.h"
+
+//  Qt
+#include <QTextStream>
+
+namespace hoot
+{
+
+XmlElement::XmlElement(const XmlObject& object, ElementIdToIdMap* idMap)
+  : _type(ElementType::Unknown),
+    _id(object.second.value("id").toString().toLong()),
+    _version(object.second.value("version").toString().toLong()),
+    _object(object),
+    _idMap(idMap),
+    _status(XmlElement::Available)
+{
+}
+
+void XmlElement::addTag(const XmlObject& tag)
+{
+  //  Make sure that the object is in fact a tag before adding it
+  if (tag.first == "tag")
+    _tags.push_back(tag);
+}
+
+QString XmlElement::toString(const QXmlStreamAttributes& attributes, long changesetId) const
+{
+  QString buffer;
+  QTextStream ts(&buffer);
+  //  Setting the codec to UTF-8 is the special sauce for foreign characters
+  ts.setCodec("UTF-8");
+  for (int i = 0; i < attributes.size(); ++i)
+  {
+    const QXmlStreamAttribute& attribute = attributes.at(i);
+    QStringRef name = attribute.name();
+    ts << name.toString() << "=\"";
+    //  Some attributes need to be overridden
+    if (name == "changeset")    //  Changeset ID is provided by the OSM API
+      ts << changesetId;
+    else if (name == "id")      //  Element ID could be different than provided in the file, could be from the API
+      ts << (_idMap ? _idMap->getNewId(_type, _id) : _id);
+    else if (name == "version") //  Versions are updated by the API
+      ts << _version;
+    else
+      ts << attribute.value().toString();
+    ts << "\" ";
+  }
+  //  Add the changeset attribute if the original data didn't include it
+  if (!attributes.hasAttribute("changeset"))
+    ts << "changeset=\"" << changesetId << "\"";
+  return ts.readAll();
+}
+
+QString& XmlElement::escapeString(QString& value) const
```

Comment:
How about `/`?

Reply:
Good call.  I don't think I saw that in my test data.

Reply:
According to [this](https://www.w3.org/TR/xml/#syntax), `/` is valid in `Character Data` but I did miss the `'` character.

Reply:
Sounds good!